File: typescript/modules/gaze_tracker.ts
/**
 * Gaze Tracker
 * Calcula o vetor de atenção do cliente.
 * Combina: Rotação da Cabeça + Posição da Íris.
 */

import type {
	GazeAnalysisResult,
	GazeStatus,
	LandmarkPoint,
	ThresholdsConfig,
} from "@type/index";

const IDX_IRIS_L = 468;
const IDX_EYE_L_CORNERS: [number, number] = [33, 133]; // Inner corner, outer corner
const IDX_NOSE = 1;
const IDX_FACE_EDGES: [number, number] = [234, 454]; // Left ear, right ear

export function createGazeTracker(config: ThresholdsConfig) {
	const maxDeviation = config.safety?.max_gaze_deviation || 100;

	/**
	 * Retorna:
	 * - isLooking (bool): True se estiver olhando para a câmera/tela.
	 * - deviation (float): Grau de desvio estimado (0 a 100)
	 * - status (str): "DIRECT", "THINKING_UP", "THINKING_DOWN", "SIDEWAY"
	 *
	 */
	function analyze(landmarks: LandmarkPoint[]): GazeAnalysisResult {
		const nose = landmarks[IDX_NOSE];
		const leftEar = landmarks[IDX_FACE_EDGES[0]];
		const rightEar = landmarks[IDX_FACE_EDGES[1]];

		if (!nose || !leftEar || !rightEar) {
			return {
				isLooking: false,
				deviation: 0.0,
				status: "ERROR",
			};
		}

		const faceWidth = rightEar.x - leftEar.x;
		if (Math.abs(faceWidth) < 1e-6) {
			return {
				isLooking: false,
				deviation: 0.0,
				status: "ERROR",
			};
		}

		const headYawRatio = (nose.x - leftEar.x) / faceWidth;
		const headYawDeviation = Math.abs(headYawRatio - 0.5) * 200; // Scale approx 0-100

		const irisL = landmarks[IDX_IRIS_L];
		const eyeLStart = landmarks[IDX_EYE_L_CORNERS[0]];
		const eyeLEnd = landmarks[IDX_EYE_L_CORNERS[1]];

		if (!irisL || !eyeLStart || !eyeLEnd) {
			return {
				isLooking: false,
				deviation: 0.0,
				status: "ERROR",
			};
		}

		const eyeLWidth = eyeLEnd.x - eyeLStart.x;

		const irisRatio =
			Math.abs(eyeLWidth) < 1e-6 ? 0.5 : (irisL.x - eyeLStart.x) / eyeLWidth;

		const irisDeviation = Math.abs(irisRatio - 0.5) * 200; // Scale approx 0-100

		const totalDeviation = headYawDeviation + irisDeviation * 0.5;

		const irisY = irisL.y;
		const eyeYCenter = (eyeLStart.y + eyeLEnd.y) / 2;
		const verticalDiff = (irisY - eyeYCenter) * 1000;

		let status: GazeStatus = "DIRECT";
		if (totalDeviation > maxDeviation) {
			if (verticalDiff < -15) {
				status = "THINKING_UP";
			} else if (verticalDiff > 15) {
				status = "THINKING_DOWN";
			} else {
				status = "SIDEWAY";
			}

			return {
				isLooking: false,
				deviation: totalDeviation,
				status,
			};
		}

		return {
			isLooking: true,
			deviation: totalDeviation,
			status: "DIRECT",
		};
	}

	return {
		analyze,
	};
}

File: python/modules/gaze_tracker.py
import numpy as np
from core.geometry_utils import GeometryUtils

class GazeTracker:
    """
    Calcula o vetor de atenção do cliente.
    Combina: Rotação da Cabeça + Posição da Íris.
    """
    def __init__(self, config):
        self.max_deviation = config['safety']['max_gaze_deviation']
        
        # Índices MP (Left=Esq na imagem, Dir real do sujeito)
        self.IDX_IRIS_L = 468
        self.IDX_IRIS_R = 473
        self.IDX_EYE_L_CORNERS = (33, 133)   # Canto interno, externo
        self.IDX_EYE_R_CORNERS = (362, 263)
        
        # Índices para estimativa de pose da cabeça (Simplificado)
        self.IDX_NOSE = 1
        self.IDX_FACE_EDGES = (234, 454) # Orelha esq, Orelha dir

    def analyze(self, landmarks, frame_width, frame_height):
        """
        Retorna:
          - is_looking (bool): True se estiver olhando para a câmera/tela.
          - deviation (float): Grau de desvio estimado (0 a 100).
          - status (str): "DIRECT", "THINKING_UP", "THINKING_DOWN", "SIDEWAY"
        """
        
        # 1. Estimativa de Pose da Cabeça (Yaw - Rotação Lateral)
        nose = landmarks[self.IDX_NOSE].x
        left_ear = landmarks[self.IDX_FACE_EDGES[0]].x
        right_ear = landmarks[self.IDX_FACE_EDGES[1]].x
        
        # Razão de simetria do nariz em relação às orelhas
        face_width = right_ear - left_ear
        if face_width == 0: return False, 0.0, "ERROR"
        
        # 0.5 = Centro. <0.5 Esquerda, >0.5 Direita
        head_yaw_ratio = (nose - left_ear) / face_width
        head_yaw_deviation = abs(head_yaw_ratio - 0.5) * 200 # Escala aprox 0-100
        
        # 2. Rastreamento de Íris (Ajuste fino)
        # Calcula onde a íris está dentro do olho
        iris_l = landmarks[self.IDX_IRIS_L].x
        eye_l_start, eye_l_end = landmarks[self.IDX_EYE_L_CORNERS[0]].x, landmarks[self.IDX_EYE_L_CORNERS[1]].x
        eye_l_width = eye_l_end - eye_l_start
        
        # Posição normalizada da íris (0.0 a 1.0 dentro do olho)
        if eye_l_width == 0: iris_ratio = 0.5
        else: iris_ratio = (iris_l - eye_l_start) / eye_l_width
        
        iris_deviation = abs(iris_ratio - 0.5) * 200 # Escala aprox 0-100

        # 3. Fusão (Score Total de Desvio)
        # Se a cabeça gira, o olho costuma compensar. Se ambos giram, desvio é alto.
        total_deviation = head_yaw_deviation + (iris_deviation * 0.5)

        # 4. Classificação Vertical (Olhar Cima/Baixo)
        # Importante para diferenciar "Pensando" (Cima/Lado) de "Tristeza" (Baixo)
        iris_y = landmarks[self.IDX_IRIS_L].y
        eye_y_center = (landmarks[self.IDX_EYE_L_CORNERS[0]].y + landmarks[self.IDX_EYE_L_CORNERS[1]].y) / 2
        vertical_diff = (iris_y - eye_y_center) * 1000
        
        status = "DIRECT"
        if total_deviation > self.max_deviation:
            if vertical_diff < -15: status = "THINKING_UP"   # Olhando p/ cima
            elif vertical_diff > 15: status = "THINKING_DOWN" # Olhando p/ baixo
            else: status = "SIDEWAY" # Olhando p/ lado
            
            return False, total_deviation, status
            
        return True, total_deviation, "DIRECT"
